mkosi/__init__.py

```diff
@@ -3797,13 +3821,117 @@ def insert_verity(
             loopdev,
             args.verity_partno,
             verity,
-            root_partition_name(args, True),
+            root_partition_name(args, "Verity"),
             gpt_root_native(args.architecture, args.usr_only).verity,
             True,
             u,
         )
 
 
+def make_verity_sig(
+    args: CommandLineArguments, root_hash: Optional[str], do_run_build_script: bool, for_cache: bool
+) -> Tuple[Optional[BinaryIO], Optional[bytes], Optional[str]]:
+
+    if do_run_build_script or args.verity != "signed":
+        return None, None, None
+    if for_cache:
+        return None, None, None
+
+    assert root_hash is not None
+
+    with complete_step("Signing verity root hash…"):
+
+        PKCS7_NOCERTS  = 0x2    # Don't include signature certificates in result
+        PKCS7_DETACHED = 0x40   # Don't include data to sign in result
+        PKCS7_BINARY   = 0x80   # Don't mangle newlines for MIME canonical format
+        PKCS7_NOATTR   = 0x100  # Don't include signature time, … in result
+
+        with open(args.secure_boot_key) as f:
+            key = crypto.load_privatekey(crypto.FILETYPE_PEM, f.read())
+
+        with open(args.secure_boot_certificate) as f:
+            certificate = crypto.load_certificate(crypto.FILETYPE_PEM, f.read())
+
+        # Note that the library returns the SHA256 digest in an
+        # uppecase format with : as byte separators. Let's convert to
+        # classic lowercase series of un-separated hex digits
+        fingerprint = certificate.digest("sha256").decode("ascii").replace(":", "").lower()
+
+        bio_in = crypto._new_mem_buf(root_hash.encode("utf-8"))
```

Comment:
Using the internal functions of an external library does introduce the possibility of breakage, since they probably don't guarantee API stability between versions.